diff --git a/channels/chan_dahdi.c b/channels/chan_dahdi.c
index ad7a8e5e54..acd76dd2c5 100644
--- a/channels/chan_dahdi.c
+++ b/channels/chan_dahdi.c
@@ -952,6 +952,7 @@ static struct dahdi_chan_conf dahdi_chan_conf_default(void)
 			.mohsuggest = "",
 			.parkinglot = "",
 			.transfertobusy = 1,
+			.realtimepulsing = 0,
 
 			.ani_info_digits = 2,
 			.ani_wink_time = 1000,
@@ -2441,6 +2442,12 @@ static enum analog_event dahdievent_to_analogevent(int event)
 	case DAHDI_EVENT_PULSE_START:
 		res = ANALOG_EVENT_PULSE_START;
 		break;
+	case DAHDI_EVENT_PULSE:
+		res = ANALOG_EVENT_PULSE;
+		break;
+	case DAHDI_EVENT_PULSE_BREAK:
+		res = ANALOG_EVENT_PULSE_BREAK;
+		break;
 	case DAHDI_EVENT_POLARITY:
 		res = ANALOG_EVENT_POLARITY;
 		break;
@@ -7463,6 +7470,10 @@ static struct ast_frame *dahdi_handle_event(struct ast_channel *ast)
 		if (!ast_channel_pbx(ast))
 			tone_zone_play_tone(p->subs[idx].dfd, -1);
 		break;
+	case DAHDI_EVENT_PULSE:
+	case DAHDI_EVENT_PULSE_BREAK:
+		/* handled in sig_analog */
+		break;
 	case DAHDI_EVENT_DIALCOMPLETE:
 		/* DAHDI has completed dialing all digits sent using DAHDI_DIAL. */
 #if defined(HAVE_PRI)
@@ -12597,6 +12608,7 @@ static struct dahdi_pvt *mkintf(int channel, const struct dahdi_chan_conf *conf,
 #endif
 		tmp->immediate = conf->chan.immediate;
 		tmp->transfertobusy = conf->chan.transfertobusy;
+		tmp->realtimepulsing = conf->chan.realtimepulsing;
 		if (chan_sig & __DAHDI_SIG_FXS) {
 			tmp->mwimonitor_fsk = conf->chan.mwimonitor_fsk;
 			tmp->mwimonitor_neon = conf->chan.mwimonitor_neon;
@@ -12930,6 +12942,7 @@ static struct dahdi_pvt *mkintf(int channel, const struct dahdi_chan_conf *conf,
 				analog_p->threewaycalling = conf->chan.threewaycalling;
 				analog_p->transfer = conf->chan.transfer;
 				analog_p->transfertobusy = conf->chan.transfertobusy;
+				analog_p->realtimepulsing = conf->chan.realtimepulsing;
 				analog_p->use_callerid = tmp->use_callerid;
 				analog_p->usedistinctiveringdetection = tmp->usedistinctiveringdetection;
 				analog_p->use_smdi = tmp->use_smdi;
@@ -18121,6 +18134,8 @@ static int process_dahdi(struct dahdi_chan_conf *confp, const char *cat, struct
 			confp->chan.immediate = ast_true(v->value);
 		} else if (!strcasecmp(v->name, "transfertobusy")) {
 			confp->chan.transfertobusy = ast_true(v->value);
+		} else if (!strcasecmp(v->name, "realtimepulsing")) {
+			confp->chan.realtimepulsing = ast_true(v->value);
 		} else if (!strcasecmp(v->name, "mwimonitor")) {
 			confp->chan.mwimonitor_neon = 0;
 			confp->chan.mwimonitor_fsk = 0;
diff --git a/channels/chan_dahdi.h b/channels/chan_dahdi.h
index de813f21bf..6681526b82 100644
--- a/channels/chan_dahdi.h
+++ b/channels/chan_dahdi.h
@@ -332,6 +332,11 @@ struct dahdi_pvt {
 	unsigned int pulse:1;
 	/*! \brief TRUE if a pulsed digit was detected. (Pulse dial phone detected) */
 	unsigned int pulsedial:1;
+	/*!
+	 * \brief TRUE if we will support real time dial pulsing.
+	 * \note Set from the "realtimepulsing" value read in from chan_dahdi.conf
+	 */
+	unsigned int realtimepulsing:1;
 	unsigned int restartpending:1;		/*!< flag to ensure counted only once for restart */
 	/*!
 	 * \brief TRUE if caller ID is restricted.
diff --git a/channels/chan_iax2.c b/channels/chan_iax2.c
index ff235af205..811fa37f0c 100644
--- a/channels/chan_iax2.c
+++ b/channels/chan_iax2.c
@@ -1385,6 +1385,7 @@ static int iax2_is_control_frame_allowed(int subtype)
 	case AST_CONTROL_TAKEOFFHOOK:
 	case AST_CONTROL_OFFHOOK:
 	case AST_CONTROL_CONGESTION:
+	case AST_CONTROL_PULSE:
 	case AST_CONTROL_FLASH:
 	case AST_CONTROL_WINK:
 	case AST_CONTROL_OPTION:
diff --git a/channels/sig_analog.c b/channels/sig_analog.c
index 64f1114fde..1f3991cf23 100644
--- a/channels/sig_analog.c
+++ b/channels/sig_analog.c
@@ -290,6 +290,12 @@ static char *analog_event2str(enum analog_event event)
 	case ANALOG_EVENT_PULSE_START:
 		res = "ANALOG_EVENT_PULSE_START";
 		break;
+	case ANALOG_EVENT_PULSE:
+		res = "ANALOG_EVENT_PULSE";
+		break;
+	case ANALOG_EVENT_PULSE_BREAK:
+		res = "ANALOG_EVENT_PULSE_BREAK";
+		break;
 	case ANALOG_EVENT_POLARITY:
 		res = "ANALOG_EVENT_POLARITY";
 		break;
@@ -2822,10 +2828,30 @@ static struct ast_frame *__analog_handle_event(struct analog_pvt *p, struct ast_
 		break;
 #endif
 	case ANALOG_EVENT_PULSE_START:
+		p->pulsemakecount = p->pulsebreakcount = 0;
 		/* Stop tone if there's a pulse start and the PBX isn't started */
 		if (!ast_channel_pbx(ast))
 			analog_play_tone(p, ANALOG_SUB_REAL, -1);
 		break;
+	case ANALOG_EVENT_PULSE:
+		if (p->realtimepulsing) {
+			ast_queue_control(p->subs[ANALOG_SUB_REAL].owner, AST_CONTROL_PULSE);
+			if (p->pulsemakecount < 9) {
+				struct timespec now = ast_tsnow();
+				p->pulsemakes[p->pulsemakecount] = now.tv_sec * 1000 + now.tv_nsec / 1000000;
+				p->pulsemakecount++;
+			}
+		}
+		break;
+	case ANALOG_EVENT_PULSE_BREAK:
+		if (p->realtimepulsing) {
+			if (p->pulsebreakcount < 9) {
+				struct timespec now = ast_tsnow();
+				p->pulsebreaks[p->pulsebreakcount] = now.tv_sec * 1000 + now.tv_nsec / 1000000;
+				p->pulsebreakcount++;
+			}
+		}
+		break;
 	case ANALOG_EVENT_DIALCOMPLETE:
 		if (p->inalarm) {
 			break;
diff --git a/channels/sig_analog.h b/channels/sig_analog.h
index 488be3662e..58a2aa4780 100644
--- a/channels/sig_analog.h
+++ b/channels/sig_analog.h
@@ -88,6 +88,8 @@ enum analog_event {
 	ANALOG_EVENT_RINGEROFF,
 	ANALOG_EVENT_HOOKCOMPLETE,
 	ANALOG_EVENT_PULSE_START,
+	ANALOG_EVENT_PULSE,
+	ANALOG_EVENT_PULSE_BREAK,
 	ANALOG_EVENT_POLARITY,
 	ANALOG_EVENT_RINGBEGIN,
 	ANALOG_EVENT_EC_DISABLED,
@@ -273,6 +275,13 @@ struct analog_pvt {
 	struct analog_dialoperation dop;
 	int onhooktime;							/*< Time the interface went on-hook. */
 	int fxsoffhookstate;					/*< TRUE if the FXS port is off-hook */
+
+	/* Used for real time dial pulsing support */
+	int pulsemakecount;
+	int pulsebreakcount;
+	int pulsemakes[9];
+	int pulsebreaks[9];
+
 	/*! \brief -1 = unknown, 0 = no messages, 1 = new messages available */
 	int msgstate;
 
@@ -291,6 +300,7 @@ struct analog_pvt {
 	unsigned int permcallwaiting:1;			/*!< TRUE if call waiting is enabled. (Configured option) */
 	unsigned int permhidecallerid:1;		/*!< Whether to hide our outgoing caller ID or not */
 	unsigned int pulse:1;
+	unsigned int realtimepulsing:1;
 	unsigned int threewaycalling:1;
 	unsigned int transfer:1;
 	unsigned int transfertobusy:1;			/*!< allow flash-transfers to busy channels */
diff --git a/funcs/func_frame_drop.c b/funcs/func_frame_drop.c
index c19c56680c..472dc6c389 100644
--- a/funcs/func_frame_drop.c
+++ b/funcs/func_frame_drop.c
@@ -72,6 +72,7 @@
 					<enum name = "TAKEOFFHOOK" />
 					<enum name = "OFFHOOK" />
 					<enum name = "CONGESTION" />
+					<enum name = "PULSE" />
 					<enum name = "FLASH" />
 					<enum name = "WINK" />
 					<enum name = "PROGRESS" />
@@ -123,6 +124,7 @@ static struct {
 	{ AST_CONTROL_TAKEOFFHOOK,   ",TAKEOFFHOOK," },
 	{ AST_CONTROL_OFFHOOK,   ",OFFHOOK," },
 	{ AST_CONTROL_CONGESTION,   ",CONGESTION," },
+	{ AST_CONTROL_PULSE,  ",PULSE," },
 	{ AST_CONTROL_FLASH,   ",FLASH," },
 	{ AST_CONTROL_WINK,   ",WINK," },
 	{ AST_CONTROL_PROGRESS,   ",PROGRESS," },
diff --git a/funcs/func_frame_trace.c b/funcs/func_frame_trace.c
index b62bae9dee..640ea8ce24 100644
--- a/funcs/func_frame_trace.c
+++ b/funcs/func_frame_trace.c
@@ -259,6 +259,9 @@ static void print_frame(struct ast_frame *frame)
 		case AST_CONTROL_CONGESTION:
 			ast_verbose("SubClass: CONGESTION\n");
 			break;
+		case AST_CONTROL_PULSE:
+			ast_verbose("SubClass: PULSE\n");
+			break;
 		case AST_CONTROL_FLASH:
 			ast_verbose("SubClass: FLASH\n");
 			break;
diff --git a/include/asterisk/frame.h b/include/asterisk/frame.h
index 6c6fafd1cf..fd0eeed399 100644
--- a/include/asterisk/frame.h
+++ b/include/asterisk/frame.h
@@ -312,6 +312,7 @@ enum ast_control_frame_type {
 	AST_CONTROL_STREAM_TOPOLOGY_REQUEST_CHANGE = 35,    /*!< Channel indication that a stream topology change has been requested */
 	AST_CONTROL_STREAM_TOPOLOGY_CHANGED = 36,           /*!< Channel indication that a stream topology change has occurred */
 	AST_CONTROL_STREAM_TOPOLOGY_SOURCE_CHANGED = 37,    /*!< Channel indication that one of the source streams has changed its source */
+	AST_CONTROL_PULSE = 38,			/*!< Dial Pulse */
 
 	/*
 	 * WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING
diff --git a/main/channel.c b/main/channel.c
index 8e1c62946b..c04fc97e93 100644
--- a/main/channel.c
+++ b/main/channel.c
@@ -3322,6 +3322,7 @@ int ast_waitfordigit_full(struct ast_channel *c, int timeout_ms, const char *bre
 				case AST_CONTROL_HOLD:
 				case AST_CONTROL_UNHOLD:
 				case AST_CONTROL_FLASH:
+				case AST_CONTROL_PULSE:
 				case -1:
 					/* Unimportant */
 					break;
@@ -4330,6 +4331,7 @@ static int attribute_const is_visible_indication(enum ast_control_frame_type con
 	case AST_CONTROL_OPTION:
 	case AST_CONTROL_WINK:
 	case AST_CONTROL_FLASH:
+	case AST_CONTROL_PULSE:
 	case AST_CONTROL_OFFHOOK:
 	case AST_CONTROL_TAKEOFFHOOK:
 	case AST_CONTROL_ANSWER:
@@ -4631,6 +4633,7 @@ static int indicate_data_internal(struct ast_channel *chan, int _condition, cons
 	case AST_CONTROL_OPTION:
 	case AST_CONTROL_WINK:
 	case AST_CONTROL_FLASH:
+	case AST_CONTROL_PULSE:
 	case AST_CONTROL_OFFHOOK:
 	case AST_CONTROL_TAKEOFFHOOK:
 	case AST_CONTROL_ANSWER:
diff --git a/main/frame.c b/main/frame.c
index 711f9e15f1..2a80921c24 100644
--- a/main/frame.c
+++ b/main/frame.c
@@ -444,6 +444,9 @@ char *ast_frame_subclass2str(struct ast_frame *f, char *subclass, size_t slen, c
 		case AST_CONTROL_CONGESTION:
 			ast_copy_string(subclass, "Congestion", slen);
 			break;
+		case AST_CONTROL_PULSE:
+			ast_copy_string(subclass, "Pulse", slen);
+			break;
 		case AST_CONTROL_FLASH:
 			ast_copy_string(subclass, "Flash", slen);
 			break;
