diff --git a/tools/objtool/check.c b/tools/objtool/check.c
index 3f7999317..4b6c8665a 100644
--- a/tools/objtool/check.c
+++ b/tools/objtool/check.c
@@ -5069,6 +5069,3 @@ int check(struct objtool_file *file)
-		if (opts.werror && warnings)
-			ret = 1;
-
 		if (opts.verbose) {
 			if (opts.werror && warnings)
 				WARN("%d warning(s) upgraded to errors", warnings);
@@ -5095,2 +5092,2 @@ int check(struct objtool_file *file)
-	return ret;
+	return opts.werror ? 1 : 0;
 }
