--- drivers/dahdi/wcte12xp/base.c.orig	2023-09-15 23:14:50.661073026 +0000
+++ drivers/dahdi/wcte12xp/base.c	2023-09-15 23:15:03.701192057 +0000
@@ -43,9 +43,15 @@
 
 #include <linux/slab.h>
 
-#include <stdbool.h>
 #include <dahdi/kernel.h>
 
+/* Linux kernel 5.16 and greater has removed user-space headers from the kernel include path */
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 16, 0)
+#include <asm/types.h>
+#else
+#include <stdbool.h>
+#endif
+
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 11, 0)
 #include <linux/sched/signal.h>
 #endif /* 4.11.0 */
